CAPÍTULO 12 - Modelos lineares
-


Esta parte do livro segue os métodos descritos no livro R for Erevyone. O raciocínio é o seguindo: como você já conhece os métodos de manipulação de dados em python usando Pandas, poderá salvar o conjunto de dados limpo caso sejá nescessário usar um método de outra linguagem de análise

12.2 Regressão linear.

O objetivo da regessão lienar é estabelecer uma relação direta entre uma  
variável de resposta (também conhecida como variável resultante ou dependente) e uma variável preditora (também conhecida como variável covariante ou independente).

In [5]:
import pandas as pd
import seaborn as sns

tips = sns.load_dataset('tips')
print(tips.head())

   total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4


Em nossa regresão linear simple, gorataríamos de ver como total_bill se relaciona com a gorjeta (tip), ou como ele a prevê.

12.2.1 Usando a stats models 

Podemos usar a biblioteca statsmodels para fazer nossa regressão linear simples. Usaremos a API formula da stasmodeis.

In [6]:
import statsmodels.formula.api as smf

Para essa regressão linear simples, usaremos a função ols, que calcula o valor dos mínimos quadrados ordinários; é um método para estimar parâmetros em uma regressão linear.

Lembre-se de que a fórmula para uma linha é y = mx + b, em que y é nossa  variável composta, x é o preditor, b é o intercepto e m é a inclinação, o parâmetro que estamos estimando.

A nptação para fórmula contém duas partes separadas por um til, ~. À esquerda do til, está a variavel de resposta, e à direita, o preditor

In [8]:
model = smf.ols(formula='tip ~ total_bill', data=tips)
results  = model.fit()

In [9]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                    tip   R-squared:                       0.457
Model:                            OLS   Adj. R-squared:                  0.454
Method:                 Least Squares   F-statistic:                     203.4
Date:                Mon, 24 Jun 2024   Prob (F-statistic):           6.69e-34
Time:                        18:20:16   Log-Likelihood:                -350.54
No. Observations:                 244   AIC:                             705.1
Df Residuals:                     242   BIC:                             712.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.9203      0.160      5.761      0.0

Nesse caso, vemos o intercept do modelo e total_bill. Esses parâmetros podem ser usados um nossa fórmula para a linha, y = (0.105)x + 0.920.   
Para interpretar esses números, dizemos que a cada incremento de uma unidade em total_bill a gorjeta aumentará (ou seja, sempre a conta aumentar em um dolar a gorjeta aumentará em 10,5 centavos)

Se quisermos apena o coeficiente podemos chamar o atributo params em results.

In [10]:
print(results.params)

Intercept     0.920270
total_bill    0.105025
dtype: float64


Dependo da sua área, você pode querer informar um intervalo de confiança, que indentifica os possíveis valores que os valores estimados poderão assumir. O intervalo de confiança inclui valores menores que [0.025 0.975].  
Também podemos obter esses valores usando o método conf_int.

In [12]:
print(results.conf_int())

                   0         1
Intercept   0.605622  1.234918
total_bill  0.090517  0.119532
